build: GCC13 compilation failures in Experimental CI

xin-zhang2 opened the pull request:
Unblock Experimental CI jobs which use Ubuntu 24.04

Fixes #12497

xin-zhang2 commented on velox/vector/fuzzer/VectorFuzzer.h:
```diff
@@ -145,11 +145,14 @@ class VectorFuzzer {
   };
 
+  // Spurious warnings in GCC 13
+  VELOX_SUPPRESS_STRINGOP_OVERFLOW_WARNING
```
@majetideepak I suppressed the stringop_overflow warning here as there is a failure in the initializaiton of opts_.
```
usr/include/c++/13/bits/stl_algobase.h:437:30: error: ‘void* __builtin_memmove(void*, const void*, long unsigned int)’ writing between 5 and 9223372036854775807 bytes into a region of size 4 overflows the destination [-Werror=stringop-overflow=]
  437 |             __builtin_memmove(__result, __first, sizeof(_Tp) * _Num);
      |             ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/x86_64-linux-gnu/c++/13/bits/c++allocator.h:33,
                 from /usr/include/c++/13/bits/allocator.h:46,
                 from /usr/include/c++/13/string:43:
In member function ‘_Tp* std::__new_allocator<_Tp>::allocate(size_type, const void*) [with _Tp = facebook::velox::fuzzer::UTF8CharList]’,
    inlined from ‘static _Tp* std::allocator_traits<std::allocator<_CharT> >::allocate(allocator_type&, size_type) [with _Tp = facebook::velox::fuzzer::UTF8CharList]’ at /usr/include/c++/13/bits/alloc_traits.h:482:28,
    inlined from ‘std::_Vector_base<_Tp, _Alloc>::pointer std::_Vector_base<_Tp, _Alloc>::_M_allocate(std::size_t) [with _Tp = facebook::velox::fuzzer::UTF8CharList; _Alloc = std::allocator<facebook::velox::fuzzer::UTF8CharList>]’ at /usr/include/c++/13/bits/stl_vector.h:381:33,
    inlined from ‘void std::_Vector_base<_Tp, _Alloc>::_M_create_storage(std::size_t) [with _Tp = facebook::velox::fuzzer::UTF8CharList; _Alloc = std::allocator<facebook::velox::fuzzer::UTF8CharList>]’ at /usr/include/c++/13/bits/stl_vector.h:398:44,
    inlined from ‘std::_Vector_base<_Tp, _Alloc>::_Vector_base(std::size_t, const allocator_type&) [with _Tp = facebook::velox::fuzzer::UTF8CharList; _Alloc = std::allocator<facebook::velox::fuzzer::UTF8CharList>]’ at /usr/include/c++/13/bits/stl_vector.h:335:26,
    inlined from ‘std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = facebook::velox::fuzzer::UTF8CharList; _Alloc = std::allocator<facebook::velox::fuzzer::UTF8CharList>]’ at /usr/include/c++/13/bits/stl_vector.h:603:61,
    inlined from ‘facebook::velox::VectorFuzzer::Options::Options(const facebook::velox::VectorFuzzer::Options&)’ at /root/velox/./velox/vector/fuzzer/VectorFuzzer.h:78:10,
    inlined from ‘facebook::velox::VectorFuzzer::VectorFuzzer(const Options&, facebook::velox::memory::MemoryPool*, size_t)’ at /root/velox/./velox/vector/fuzzer/VectorFuzzer.h:153:9,
    inlined from ‘ExchangeFuzzer::ExchangeFuzzer()’ at /root/velox/velox/exec/tests/ExchangeFuzzer.cpp:96:22:
```
The code in VectorFuzzer.h:153 is `: opts_(options), pool_(pool), rng_(seed) {}`.

This failure should be caused by a bug in GCC 13 and it only occurs in the release build.
I also found some similiar issues online like [114758](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114758) and [111415](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=111415).

How do you think we should resolve this issue?

czentgr replied:
Is it known which versions have the bug? You could add a define to activate only of the offending versions of gcc are used to compile.
This is only ion the fuzzer so I don't think it's a big issue adding the warning suppression here.
We've had a few problems like that. So if you add compiler and version specific check around it it should be ok.

xin-zhang2 replied:
@czentgr 
It looks the issue was introduced in GCC 12.4, and has not been fixed yet in the latest 14.2.
https://godbolt.org/z/d7PK58hxT

I didn't reproduce it with GCC trunk, so it is expected to be fixed in the next GCC release.
I have added the condition to check the compiler and its version.

xin-zhang2 commented on velox/connectors/tpch/TpchConnector.cpp:
```diff
@@ -87,12 +87,7 @@ TpchDataSource::TpchDataSource(
-        "for '{}' on table '{}'",
-        handle->name(),
-        toTableName(tpchTable_));
+    VELOX_CHECK_NOT_NULL(handle, "ColumnHandle can't be null.");
```
When handle is nullptr, `handle->name()` in the previous message will cause a null pointer dereference.
```
/root/velox/velox/connectors/tpch/TpchConnector.cpp:94:21: error: ‘this’ pointer is null [-Werror=nonnull]
   94 |         handle->name(),
      |         ~~~~~~~~~~~~^~
/root/velox/./velox/common/base/Exceptions.h:169:60: note: in definition of macro ‘_VELOX_THROW_IMPL’
  169 |     auto message = ::facebook::velox::detail::errorMessage(__VA_ARGS__); \
      |                                                            ^~~~~~~~~~~
/root/velox/./velox/common/base/Exceptions.h:196:3: note: in expansion of macro ‘_VELOX_CHECK_AND_THROW_IMPL’
  196 |   _VELOX_CHECK_AND_THROW_IMPL(                                      \
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/root/velox/./velox/common/base/Exceptions.h:271:32: note: in expansion of macro ‘_VELOX_CHECK_IMPL’
  271 | #define VELOX_CHECK(expr, ...) _VELOX_CHECK_IMPL(expr, #expr, ##__VA_ARGS__)
      |                                ^~~~~~~~~~~~~~~~~
/root/velox/./velox/common/base/Exceptions.h:279:38: note: in expansion of macro ‘VELOX_CHECK’
  279 | #define VELOX_CHECK_NOT_NULL(e, ...) VELOX_CHECK(e != nullptr, ##__VA_ARGS__)
      |                                      ^~~~~~~~~~~
/root/velox/velox/connectors/tpch/TpchConnector.cpp:90:5: note: in expansion of macro ‘VELOX_CHECK_NOT_NULL’
   90 |     VELOX_CHECK_NOT_NULL(
      |     ^~~~~~~~~~~~~~~~~~~~
```

majetideepak replied:
`handle->name()` must be replaced with `it->second->name()`.
But nice find by GCC 13.

xin-zhang2 replied:
Updated.

xin-zhang2 commented on velox/dwio/dwrf/common/Compression.h:
```diff
@@ -87,7 +87,7 @@ inline std::unique_ptr<dwio::common::BufferedOutputStream> createCompressor(
 inline CompressionOptions getDwrfOrcDecompressionOptions(
     common::CompressionKind kind) {
-  CompressionOptions options;
+  CompressionOptions options{};
```
Initialize `options` explicitly to avoid using it uninitialized.
```
/root/velox/./velox/dwio/dwrf/common/Compression.h:99:10: error: ‘options’ is used uninitialized [-Werror=uninitialized]
   99 |   return options;
      |          ^~~~~~~
/root/velox/./velox/dwio/dwrf/common/Compression.h: In member function ‘virtual void TestSeek_uncompressed_Test::TestBody()’:
/root/velox/./velox/dwio/dwrf/common/Compression.h:90:22: note: ‘options’ declared here
   90 |   CompressionOptions options;
      |                      ^~~~~~~
```

majetideepak replied:
Let's initialize the fields in the `struct CompressionOptions`

xin-zhang2 replied:
Initialized the fields.
There is no default value for `CompressionOptions.format`, so I left it empty initialized, which results in `format.zlib.windowBits` and `format.zlib.compressionLevel` being initialized to 0.
As for `CompressionOptions.compressionThreshold`, it is currently used only by `DwrfOrcCompression`, where its value is configured via `orc.compression.threshold`, with a default of 256. However, since `CompressionOptions` is part of the common package, I believe it would be more appropriate to initialize `compressionThreshold` to 0.
Does the approach seem reasonable?

xin-zhang2 commented on velox/expression/tests/ExpressionRunner.cpp:
```diff
@@ -230,10 +230,7 @@ void ExpressionRunner::run(
-          inputVector,
-          "Input vector is not a RowVector: {}",
-          inputVector->toString());
+      VELOX_CHECK_NOT_NULL(inputVector, "Input vector can't be null.");
```
When inputVector is nullptr, `inputVector->toString()` in the previous message will cause a null pointer dereference.

majetideepak replied:
```
auto vector = restoreVectorFromFile(inputPathsList[i].c_str(), deserializerPool.get());
auto inputVector = std::dynamic_pointer_cast<RowVector>(vector);

use vector->toString() in the null check argument.
```

xin-zhang2 replied:
Updated.

majetideepak commented on velox/dwio/common/compression/Compression.h:
```diff
@@ -84,9 +84,9 @@ struct CompressionOptions {
       bool isHadoopFrameFormat;
     } lz4_lzo;
-  } format;
+  } format{};
```
We must default each field to 0.

majetideepak replied:
`{}` is not required

xin-zhang2 replied:
Will default all scalar fields to 0.
We need `{}` either after `zlib` or `format` to make `format` default intialized, otherwise we have to explicitly implement a constructor for the initialization of `format`.

majetideepak replied:
If the fields are initialized, we should be good no?

xin-zhang2 replied:
yes, both the union field `format` and scalar field `compressionThreshold` should be initialized.

If all scalar fields default to 0,  the simplest way is to only add `{}` after `format`. It will execute a [value initialization](https://en.cppreference.com/w/cpp/language/value_initialization) that zero-intialize the union, thus no initializers required for structs in this union, and the code is like
```
struct CompressionOptions {
  union Format {
    struct {
      int windowBits;
      int32_t compressionLevel;
    } zlib;

    struct {
      int32_t compressionLevel;
    } zstd;

    struct {
      bool isHadoopFrameFormat;
    } lz4_lzo;
  } format{};

  uint32_t compressionThreshold{0};
};
```

If we don't want zero-initialize `format`, it can be [default initialized](https://en.cppreference.com/w/cpp/language/default_initialization) without using `{}` after `format`. In this case,  exactly one member of the union should be initialized. The code is like
```
struct CompressionOptions {
  union Format {
    struct {
      int windowBits{0};
      int32_t compressionLevel{0};
    } zlib {};

    struct {
      int32_t compressionLevel;
    } zstd;

    struct {
      bool isHadoopFrameFormat;
    } lz4_lzo;
  } format;

  uint32_t compressionThreshold{0};
};
```

majetideepak commented:
@xin-zhang2 The experimental CI job that uses GCC 13 needs to be triggered to test this. Can you add a comment somewhere in .github/workflows/experimental.yml  to trigger the CI? Thanks.

xin-zhang2 commented:
@majetideepak I've added a comment to trigger the CI, but it looks the Experimental Fuzzer Job didn't pick my code.
For example, the error message in the job is 
```
/home/runner/work/velox/velox/velox/velox/vector/tests/VectorMakerTest.cpp:477:65: error: converting to ‘std::optional<std::vector<std::optional<int>, std::allocator<std::optional<int> > > >’ from initializer list would use explicit constructor ‘constexpr std::optional<_Tp>::optional(std::in_place_t, _Args&& ...) [with _Args = {}; typename std::enable_if<__and_v<std::is_constructible<_Tp, _Args ...> >, bool>::type <anonymous> = false; _Tp = std::vector<std::optional<int>, std::allocator<std::optional<int> > >]’
  477 |   VectorPtr expectedVector = maker_.arrayVectorNullable<int32_t>({
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
  478 |       {{1, 2}},
      |       ~~~~~~~~~                                                  
  479 |       {{2, 3, 4}},
      |       ~~~~~~~~~~~~                                               
  480 |       {{std::nullopt, 7}},
      |       ~~~~~~~~~~~~~~~~~~~~                                       
  481 |       {{1, 3, 7, 9}},
      |       ~~~~~~~~~~~~~~~                                            
  482 |       {{}},
      |       ~~~~~                                                      
  483 |       {{std::nullopt}},
      |       ~~~~~~~~~~~~~~~~~                                          
  484 |       {{1, 2, std::nullopt}},
      |       ~~~~~~~~~~~~~~~~~~~~~~~                                    
  485 |       {{}},
      |       ~~~~~                                                      
  486 |       std::nullopt,
      |       ~~~~~~~~~~~~~                                              
  487 |       {{1, 2, 3}},
      |       ~~~~~~~~~~~~                                               
  488 |       {{}},
      |       ~~~~~                                                      
  489 |       {{4, 5}},
      |       ~~~~~~~~~                                                  
  490 |   });
      |   ~~          
```

while I have already made the following change in my PR.
![image](https://github.com/user-attachments/assets/f37f6d05-f49f-482c-b0bc-f2fe906342cf)

Any idea on this?

majetideepak commented:
@xin-zhang2 CI is checking out the main branch. Are you able to test this locally? If yes, we an undo the comment.

xin-zhang2 commented:
@majetideepak  Yes, the build runs successfully on my local ubuntu 24.04 enviroment.
I have reverted the comment commit.

majetideepak commented:
@xin-zhang2 thanks for this fix. Meta will import and merge this.

Yuhta commented on velox/connectors/tpch/TpchConnector.cpp:
```diff
@@ -91,7 +91,7 @@ TpchDataSource::TpchDataSource(
         "ColumnHandle must be an instance of TpchColumnHandle "
         "for '{}' on table '{}'",
-        handle->name(),
+        it->second->name(),
```
Why do we need to change this?  Looks like a readability regression to me

xin-zhang2 replied:
`handle` is cast from `it->second` using
```
auto handle = std::dynamic_pointer_cast<TpchColumnHandle>(it->second);
```
If the cast fails, `handle` will be nullptr, and `handle->name()` will cause a nullptr dereference.

Yuhta commented on velox/dwio/common/compression/Compression.h:
```diff
@@ -84,9 +84,9 @@ struct CompressionOptions {
       bool isHadoopFrameFormat;
     } lz4_lzo;
-  } format;
+  } format{};
```
I honestly don't see the value zeroing out these stuff.  People need to initialize them when they want to use them.  In this case it's fine to pay the extra initialization cost, but that's not a general rule we should follow in this code base.

xin-zhang2 replied:
CompressionOptions is used uninitialized in code like velox/dwio/dwrf/common/Compression.h
```
inline CompressionOptions getDwrfOrcDecompressionOptions(
    common::CompressionKind kind) {
  CompressionOptions options;
  if (kind == common::CompressionKind_ZLIB ||
      kind == common::CompressionKind_GZIP) {
    options.format.zlib.windowBits = Compressor::DWRF_ORC_ZLIB_WINDOW_BITS;
  } else if (
      kind == common::CompressionKind_LZ4 ||
      kind == common::CompressionKind_LZO) {
    options.format.lz4_lzo.isHadoopFrameFormat = false;
  }
  return options;
}
```
We made this change to default initialize it to avoid the following compliation errors.
```
    inlined from ‘std::unique_ptr<facebook::velox::dwio::common::SeekableInputStream> TestSeek::createTestDecompressor(facebook::velox::common::CompressionKind, std::unique_ptr<facebook::velox::dwio::common::SeekableInputStream>, uint64_t)’ at /root/velox/velox/dwio/dwrf/test/TestDecompression.cpp:873:73,
    inlined from ‘virtual void TestSeek_uncompressed_Test::TestBody()’ at /root/velox/velox/dwio/dwrf/test/TestDecompression.cpp:962:20:
/root/velox/./velox/dwio/dwrf/common/Compression.h:99:10: error: ‘options’ is used uninitialized [-Werror=uninitialized]
   99 |   return options;
      |          ^~~~~~~
/root/velox/./velox/dwio/dwrf/common/Compression.h: In member function ‘virtual void TestSeek_uncompressed_Test::TestBody()’:
/root/velox/./velox/dwio/dwrf/common/Compression.h:90:22: note: ‘options’ declared here
   90 |   CompressionOptions options;
      |                      ^~~~~~~
cc1plus: all warnings being treated as errors
```

majetideepak replied:
> I honestly don't see the value zeroing out these stuff.  
but that's not a general rule we should follow in this code base.

Initializing variables is one of the basic security principles. Velox code would not pass most security scans if this is not done. We see this a lot in IBM code scans. How does Meta handle security scans?

majetideepak replied:
I have to clarify that without the defaults, there is a good chance users might not initialize variables properly.

Yuhta commented on velox/expression/tests/ExpressionRunner.cpp:
```diff
@@ -227,13 +227,13 @@ void ExpressionRunner::run(
           inputVector,
           "Input vector is not a RowVector: {}",
-          inputVector->toString());
+          vector->toString());
```
Again why we cannot call `toString` from `inputVector`?  Looks like a bug in compiler.

xin-zhang2 replied:
The reason is same as `handle` above. 
If `inputVector` is nullptr, `toString` will cause nullptr dereference.

Yuhta commented on velox/expression/tests/GenericViewTest.cpp:
```diff
@@ -38,7 +38,7 @@ class GenericViewTest : public functions::test::FunctionBaseTest {
 
   array_data_t arrayData1 = {
-      {{}},
+      std::make_optional<std::vector<std::optional<int64_t>>>({}),
```
This `make_optional` thing looks really like a readability regression, can we avoid doing it?

xin-zhang2 replied:
Passing an initializer list to the std::optional<T> constructor in GCC13 will fail because the constructor is explicit declared. The error (not warning) is 
```
/root/velox/velox/expression/tests/GenericViewTest.cpp:56:3: error: converting to ‘std::optional<std::vector<std::optional<long int> > >’ from initializer list would use explicit constructor ‘constexpr std::optional<_Tp>::optional(std::in_place_t, _Args&& ...) [with _Args = {}; typename std::enable_if<__and_v<std::is_constructible<_Tp, _Args ...> >, bool>::type <anonymous> = false; _Tp = std::vector<std::optional<long int> >]’
   56 |   };
      |   ^
/root/velox/velox/expression/tests/GenericViewTest.cpp:56:3: error: converting to ‘std::in_place_t’ from initializer list would use explicit constructor ‘constexpr std::in_place_t::in_place_t()’
```
That's why we have to explicitly construct it with `std::make_optional`.

Tested it in https://godbolt.org/z/qxK6b5jbY and the error starts from GCC13.1.

Yuhta replied:
This is apparently a compiler bug that it should match `optional(U&& value)` signature but it tries to match the `in_place_t` tag version instead.  Clang is not suffering from this bug and it should be fixed in GCC.

Yuhta replied:
In the meantime, you can do this to mitigate the readability regression while we waiting for the GCC fix (put this in `common/testutil` as it should be used in only tests):
```c++
// Workaround for GCC to avoid matching optional(std::in_place_t, ...)
struct OptionalEmptyT {
    template <typename T>
    operator std::optional<std::vector<T>>() const {
        return std::make_optional<std::vector<T>>({});
    }
};

constexpr OptionalEmptyT optionalEmpty;

int main() {
    std::vector<std::optional<std::vector<int>>> v = {optionalEmpty};
    return 0;
}
```

majetideepak commented:
@xin-zhang2 is this ready for review?

xin-zhang2 commented:
@majetideepak Yes, I have updated this PR based on @Yuhta 's review.
1. Introduced `OptionalEmptyT` in `common/testutil` to workaround the GCC `std::optional<T>` constructor signature mismatch.
2. Initialize `CompressionOptions` at use site instead of definiton.

@Yuhta Can you help review the changes? Thanks!

Yuhta commented:
@xin-zhang2 Can you rebase to the newest commit?  Seems there is some conflicts

xin-zhang2 commented:
@Yuhta rebased to the latest commit, and didn't encounter conflict issues. Can you try re-import it? Thanks.
